code/WorkInProgress/nadir_antenna.dm

```diff
@@ -165,34 +210,285 @@ var/global/obj/machinery/communications_dish/transception/transception_array
+		if (use_standard_failsafe) //slightly over failsafe values in each case so it doesn't just turn right back on again
 			combined_cost = (0.4 * C.maxcharge) + ARRAY_STARTCOST
 		else
 			combined_cost = (0.1 * C.maxcharge) + ARRAY_STARTCOST
 		if (C.charge > combined_cost)
 			playsound(src.loc, 'sound/machines/shieldgen_startup.ogg', 50, 1)
 			src.primed = TRUE
-			src.failsafe_active = FALSE
 			src.UpdateIcon()
 			. = TRUE
 		return
 
-	ex_act(severity) //tbi: damage and repair
+	proc/charge_intcap()
+		if(src.intcap && src.intcap_draw_rate > 0)
+			if(src.intcap.rigged)
+				intcap_failure()
+				return
+
+			var/datum/powernet/powernet = src.get_direct_powernet()
+
+			//if we're not charging a cell yet, figure out what we'd be billing the powernet if we were
+			var/total_load = src.intcap_charging ? powernet.load : powernet.load + src.intcap_draw_rate
+
+			if(powernet.avail - total_load >= src.grid_surplus_threshold) //netexcess exists but... isn't ever actually set up?
+				src.intcap_charging = TRUE
+				if(src.intcap.charge < src.intcap.maxcharge)
+					var/yield_to_cell = src.intcap_draw_rate * CELLRATE
+					var/final_draw = src.intcap_draw_rate
+					//this bit is so you don't spend more on charge than you actually need to charge the cell
+					if(intcap.charge + yield_to_cell > src.intcap.maxcharge)
+						yield_to_cell = src.intcap.maxcharge - src.intcap.charge
+						final_draw = yield_to_cell * 500
+					src.intcap.give(yield_to_cell)
+					powernet.newload += final_draw
+					var/area/arrayarea = get_area(src) //gotta let the grid know!
+					arrayarea.use_power(final_draw,EQUIP)
+			else
+				src.intcap_charging = FALSE
+		else
+			src.intcap_charging = FALSE
+
+	///Layer for using internal capacitor; separated to intercept rigged cells and handle with custom damage behavior
+	proc/use_intcap(var/use_amount)
+		. = TRUE
+		if(src.intcap.rigged)
+			intcap_failure()
+			. = FALSE
+		else
+			src.intcap.use(use_amount)
+
+	///Sabotaged cells, instead of blowing out the turf, will blow out the associated microvoltage cabinet and bring the array offline
+	proc/intcap_failure()
+		src.intcap.rigged = FALSE
+		src.intcap_charging = FALSE
+		src.status |= BROKEN
+		src.primed = FALSE
+		src.intcap_door_open = FALSE
+		src.repair_status = 0
+		src.UpdateIcon()
+		if (intcap.rigger)
+			message_admins("[key_name(intcap.rigger)]'s rigged cell damaged the transception array at [log_loc(src)].")
+			logTheThing(LOG_COMBAT, intcap.rigger, "'s rigged cell damaged the transception array at [log_loc(src)].")
+
+		src.visible_message("<span class='alert'><b>[src]'s internal capacitor compartment explodes!</b></span>")
+
+		for(var/client/C in clients)
+			playsound(C.mob, 'sound/effects/explosionfar.ogg', 35, 0)
+
+		var/epicenter = get_turf(src)
+		playsound(epicenter, "explosion", 90, 1)
+		//this doesn't actually explode because turf safe explosions don't respect "space" (ocean) turfs, and I'd like surroundings not punctured
```

Comment:
Maybe just do a big fireflash? Seems weird that this plays the Big Explosion Noise but if I'm standing next to it it'll be unharmed

Reply:
Fireflashes *also* don't work on /turf/space. The best way to address this would probably be to make turf-safe explosions respect ocean turfs, but that (and at that point the reintroduction of an explosion to this) would be separate PRs.

Reply:
Seems reasonable